### PR TITLE
Support Embroider with staticComponents for full tree-shaking

### LINKED ISSUE #1206: Support staticComponents with Embroider
Currently we use string properties defined on a component class and the `{{component}}` helper to refer to other used components like here: https://github.com/kaliber5/ember-bootstrap/blob/master/addon/components/bs-form/element.js#L816, or even "worse" this: https://github.com/kaliber5/ember-bootstrap/blob/master/addon/components/bs-form/element.js#L797-L808.

This is done so we can override components from the caller side, so the actually used components are not hardcoded in the template. This pattern however makes it basically impossible to statically analyze component usage, therefore the more aggressive `staticComponents` of Embroider to tree-shake unused components would certainly not work currently.

We should support that in the future. Which AFAIU means getting rid of component strings defined on the class. Importing a component class and passing that to the component helper *should* work.

Related:
* https://github.com/emberjs/ember.js/issues/19099
* https://github.com/kaliber5/ember-bootstrap/pull/1125

### PR DESCRIPTION
**Pull request by simonihmig**
Replaces #1125, fixes #1206 

~Forms haven't been updated yet, other than that all components pass locally with strict embroider settings.~ Only locally, because I needed to patch some thing, which still have to be fixed upstream or they are already but nor released yet:

* https://github.com/embroider-build/embroider/issues/539
* ~https://github.com/embroider-build/embroider/pull/536~
* ~https://github.com/embroider-build/embroider/pull/521~

@jelhan As you can see, this adds quite a bit of boilerplate, but this seems the only way for now to make it work. See https://github.com/embroider-build/embroider/issues/487

### CODE REVIEW

File: addon/components/bs-navbar.hbs
Comment (jelhan) [inline addon/components/bs-navbar.hbs:13]: Leftover from debugging.

```suggestion
```

File: tests/integration/components/bs-modal-test.js
Comment (jelhan) [inline tests/integration/components/bs-modal-test.js:60]: Seems as we have missed that one when refactoring to angle bracket component invocation. As we are touching it now, let's fix it:

```suggestion
        <TestComponent>
```

File: tests/integration/components/bs-modal-test.js
Comment (jelhan) [inline tests/integration/components/bs-modal-test.js:63]:
```suggestion
        </TestComponent>
```

File: tests/integration/components/bs-tooltip-test.js
Comment (jelhan) [inline tests/integration/components/bs-tooltip-test.js:144]:
```suggestion
      hbs`<TestComponent id="target"><BsTooltip @title="Dummy" @fade={{false}} @triggerElement="parentView" /></TestComponent>`
```

**Review by jelhan — COMMENTED** [review]
Thanks a lot for working on that one! :clap: I'm sorry that I needed so long for the review.

Sometimes `(bs-default @passedIncomponent (component "default-component"))` is wrapped in another `{{component}}` helper even if no arguments are passed to it. An example is `addon/components/bs-dropdown/menu.hbs`, which has:

```hbs
      {{yield
        (hash
          item=(component (bs-default @itemComponent (component "bs-dropdown/menu/item")))
          link-to=(component (bs-default @linkToComponent (component "bs-dropdown/menu/link-to")))
          linkTo=(component (bs-default @linkToComponent (component "bs-dropdown/menu/link-to")))
          divider=(component (bs-default @dividerComponent (component "bs-dropdown/menu/divider")))
        )
      }}
```

This pattern is not used everywhere. An example for the opposite is yielded `title` in `addon/components/bs-modal/header.hbs`:

```hbs
         title=(bs-default @titleComponent (component "bs-modal/header/title")) 
```

I would prefer to only use a wrapping `{{component}}` helper if required to apply additional arguments.

> Forms haven't been updated yet

Any specific reason you left them out for know? Or just due to limited time?

> As you can see, this adds quite a bit of boilerplate, but this seems the only way for now to make it work.

To be honest I don't see it as a big issue. You are talking about requiring `{{bs-default}}`, aren't you? Some more code in the template but one property less in the corresponding JavaScript component. And I like that I can see the default component yielded in the template directly.

**Comment by simonihmig** [comment]
> I would prefer to only use a wrapping {{component}} helper if required to apply additional arguments.

Absolutely. This was probably lazy copy pasting, will fix it!

> Any specific reason you left them out for know? Or just due to limited time?

Yes, the latter. And the fact that I expect some trouble with the way we dynamically determine the control component [here](https://github.com/kaliber5/ember-bootstrap/blob/master/addon/components/bs-form/element.js#L796-L805)

> To be honest I don't see it as a big issue. You are talking about requiring {{bs-default}}, aren't you?

Yes. And that double wrapping of the component helper. And we have some duplicated default values, like [here](https://github.com/kaliber5/ember-bootstrap/pull/1251/files#diff-aa0d518fb94f8351df46f7626de5a02eR19-R51) vs. the single source of truth we had before when defined on the component class.

But I was more worried that you might object here! :wink: Although I see the drawbacks, I still think the benefit of enabling tree shaking outweighs this.

File: addon/components/bs-navbar.hbs
Comment (simonihmig) [reply addon/components/bs-navbar.hbs:7]: Hm, good catch! Idk honestly, maybe I just tried something out and forget to revert/cleanup. Will fix!

**Comment by jelhan** [comment]
>> Any specific reason you left them out for know? Or just due to limited time?
> 
> Yes, the latter. And the fact that I expect some trouble with the way we dynamically determine the control component here

I see. I'm totally fine with doing that in a follow-up merge request. It's your call.

> And we have some duplicated default values, like here vs. the single source of truth we had before when defined on the component class.

Being able to see it in the template directly outweights the duplication for me. Being able to be treeshake with Embroider is the icing on the cake. 🙃

File: addon/components/bs-navbar.hbs
Comment (simonihmig) [reply addon/components/bs-navbar.hbs:7]: Done

**Comment by simonihmig** [comment]
Also refactored all form related components. Everything passing now locally, still waiting for a release of https://github.com/embroider-build/embroider/pull/540

**Review by jelhan — APPROVED** [review]
Awesome work! :clap: 

One Bootstrap 3 test is failing. Have had a quick look but don't see how it could be related.

```
not ok 228 Chrome 85.0 - [69 ms] - Integration | Component | bs-form/element: adjusts validation icon position if there is an input group [BS3]
    ---
        actual: >
            null
        stack: >
            TypeError: Cannot read property 'style' of null
                at Object._callee25$ (http://localhost:7357/assets/tests.js:11132:89)
                at tryCatch (http://localhost:7357/assets/vendor.js:413:40)
                at Generator.invoke [as _invoke] (http://localhost:7357/assets/vendor.js:687:22)
                at Generator.prototype.<computed> [as next] (http://localhost:7357/assets/vendor.js:446:21)
                at asyncGeneratorStep (http://localhost:7357/assets/tests.js:9966:105)
                at _next (http://localhost:7357/assets/tests.js:9968:196)
        message: >
            Promise rejected during "adjusts validation icon position if there is an input group [BS3]": Cannot read property 'style' of null
        negative: >
            false
        browser log: |
    ...
```

**Comment by simonihmig** [comment]
> Have had a quick look but don't see how it could be related.

It was, but took my a while to debug this. Should be fixed now.

**Comment by simonihmig** [comment]
With the newly released embroider v0.26.0 this is now passing! :tada:

**Pull request merged**
